volume/drivers/pwx/connection.go

```diff
@@ -0,0 +1,210 @@
+	EnableTLSEnv:                "PX_ENABLE_TLS",
+	NamespaceNameEnv:            "PX_NAMESPACE",
+	ServiceNameEnv:              "PX_SERVICE_NAME",
+	CaCertSecretEnv:             "PX_CA_CERT_SECRET",
+	CaCertSecretKeyEnv:          "PX_CA_CERT_SECRET_KEY",
+}
+
+func NewConnectionParamsBuilder(ops core.Ops, params *ConnectionParamsBuilderConfig) *ConnectionParamsBuilder {
+	if params == nil {
+		params = &ConnectionParamsBuilderConfig{}
+	}
+
+	if params.DefaultServiceName == "" {
+		params.DefaultServiceName = DefaultParams.DefaultServiceName
+	}
+
+	if params.DefaultServiceNamespaceName == "" {
+		params.DefaultServiceNamespaceName = DefaultParams.DefaultServiceNamespaceName
+	}
+
+	if params.DefaultRestPortName == "" {
+		params.DefaultRestPortName = DefaultParams.DefaultRestPortName
+	}
+
+	if params.DefaultSDKPortName == "" {
+		params.DefaultSDKPortName = DefaultParams.DefaultSDKPortName
+	}
+
+	if params.EnableTLSEnv == "" {
+		params.EnableTLSEnv = DefaultParams.EnableTLSEnv
+	}
+
+	if params.NamespaceNameEnv == "" {
+		params.NamespaceNameEnv = DefaultParams.NamespaceNameEnv
+	}
+
+	if params.ServiceNameEnv == "" {
+		params.ServiceNameEnv = DefaultParams.ServiceNameEnv
+	}
+
+	if params.CaCertSecretEnv == "" {
+		params.CaCertSecretEnv = DefaultParams.CaCertSecretEnv
+	}
+
+	if params.CaCertSecretKeyEnv == "" {
+		params.CaCertSecretKeyEnv = DefaultParams.CaCertSecretKeyEnv
+	}
+
+	return &ConnectionParamsBuilder{
+		Config:  params,
+		kubeOps: ops,
+	}
+}
+
+// BuildClientsEndpoints returns two endpoints for PX MGMT API and gRPC SDK/iSDL API
+func (cpb *ConnectionParamsBuilder) BuildClientsEndpoints() (pxMgmtEndpoint string, sdkEndpoint string, err error) {
+	var endpoint string
+
+	// Check if service name and namespace is provided
+	// as environment variables
+	serviceName := getPxService(cpb.Config.ServiceNameEnv, cpb.Config.DefaultServiceName)
+	ns := getPxNamespace(cpb.Config.NamespaceNameEnv, cpb.Config.DefaultServiceNamespaceName)
+
+	svc, err := cpb.kubeOps.GetService(serviceName, ns)
+	if err == nil {
+		endpoint = svc.Spec.ClusterIP
+	} else {
+		return "", "", fmt.Errorf("failed to get k8s service spec: %v", err)
+	}
+
+	if len(endpoint) == 0 {
+		return "", "", fmt.Errorf("failed to get endpoint for portworx volume driver")
+	}
+
+	var restPort int
+	var sdkPort int
+
+	// Get the ports from service
+	for _, svcPort := range svc.Spec.Ports {
+		if svcPort.Name == cpb.Config.DefaultSDKPortName &&
+			svcPort.Port != 0 {
+			sdkPort = int(svcPort.Port)
+		} else if svcPort.Name == cpb.Config.DefaultRestPortName &&
+			svcPort.Port != 0 {
+			restPort = int(svcPort.Port)
+		}
+	}
+
+	// check if the ports were parsed
+	if sdkPort == 0 || restPort == 0 {
+		err := fmt.Errorf("%s in %s namespace has been not updated to the latest spec. "+
+			"%s and/or %s ports are missing. Follow "+
+			"https://docs.portworx.com/portworx-install-with-kubernetes/operate-and-maintain-on-kubernetes/upgrade/"+
```

Comment:
We should not be adding links to the docs, since the docs can move after this code is released. The link should be removed.

Reply:
Got it. That error was copied from Porx repo. 
Fixed.